### ISSUE TITLE
Unexpectedly Low BUSCO Scores in TOGA Alignment: Potential Parameter Issue with lastz?

### ISSUE THREAD

**Issue by ZhaoHang-bio** [hillerlab/TOGA#123]
Hi,

I've been testing the TOGA software using the genomes of two plants from the same genus. My parameters for the software are as follows:

`toga.py ${chain_file} ${bed_files} ${refgenome} ${quegenome} --kt --pn toga_out -i ${isoform} --cb 3,5 --cjn 10 --ms --nextflow_config_dir ${path_to_nextflow_config_dir}`

The BUSCO results for the reference genome's encoded proteins is C:93.0% [S:40.8%, D:52.2%], F:4.6%, M:2.4%. However, the alignment results from TOGA are quite different, showing C:51.3% [S:33.0%, D:18.3%], F:4.8%, M:43.9%. These results are significantly lower than I expected. Could this discrepancy be due to potentially lenient parameter settings in lastz, leading to incomplete results?

My parameters for lastz are as follows:

`lastz_32 $ref.genome[multiple] $query.genome[multiple] --masking=10 --ambiguous=iupac --chain --step=20 --rdotplot=ref.genome_vs_query.genome.dotplot --notransition H=2000 Y=3400 L=6000 K=2200 --format=axt --gfextend > ref.genome_vs_query.genome.axt`

I'm looking forward to your insights or suggestions on whether this is a normal occurrence or if there's a potential issue with the parameter settings.

Best regards,
Hang

**Comment by MichaelHiller** [comment]
Hi Hang, 

reduced lastz sensitivity could be a problem (we use more sensitive parameters), but I think this is unlikely to explain the BUSCO scores. 
Can you pls run BUSCO on the input annotation to see how complete that is?

**Comment by ZhaoHang-bio** [comment]
Hi,

Thank you for your response. 

I've already assessed the completeness of my input annotation using BUSCO, which showed C:93.0% [S:40.8%, D:52.2%], F:4.6%, M:2.4%. This includes 174,192 alternative splicing transcripts and 47,273 protein-coding genes.

However, upon further analysis of the TOGA results, I've encountered some discrepancies. The query proteins result was conducted on the 'prot.fasta' file from the output dir. The extracted query proteins completeness is 51.3%, while it's 54.3% for the reference proteins. Moreover, there seems to be a mismatch in the gene counts. The reference protein-coding genes identified are only 21,377, whereas I expected 47,273 based on the initial data.

This discrepancy is quite puzzling. Could it indicate that the software did not run to completion, or might there be another issue affecting the results?

Looking forward to your insights.

Best regards,
Hang

**Comment by MichaelHiller** [comment]
47000 genes tells me that you did not assign transcripts to genes. This is not a problem for annotation (query_annotation.bed) but the orthology assignment would be often wrong. 

Wrt BUSCO: Pls do a 
grep GENE loss_summ_data.tsv | cut -f3 | sort | uniq -c
to get a breakdown of the gene classifications. Maybe there are many Lost or missing transcripts. 
Do you think assembly fragmentation or low base accuracy can be a problem?

**Comment by ZhaoHang-bio** [comment]
Could you please clarify how to properly assign transcripts to genes? I noticed that Isoform data input is required, so I did not reduce the annotation file to only the longest transcript. Could this approach be problematic?

Regarding the command line you provided, my results were:
```
      4 I
     23 L
  13595 M
      1 UL
```
It appears there are indeed many missing genes.

As for the possibility of assembly fragmentation or low base accuracy you mentioned, the genome I'm working with was assembled using HiFi sequencing technology, with an N50 greater than 6Mb, and it's assembled at the chromosomal level. I believe these factors should mitigate the concerns about fragmentation or base accuracy.

I am sending you the TOGA run log to see if it might contain any clues to these issues.
[F-GA-01.toga.log.zip](https://github.com/hillerlab/TOGA/files/13427922/F-GA-01.toga.log.zip)

Best regards,
Hang

**Comment by MichaelHiller** [comment]
The 13k missing genes explain the low BUSCO. 
Which species are you running TOGA on? Are they within an evolutionary distance that should work with TOGA or are they very distant (the latter is a typical reason why gene can be missing)?

**Comment by ZhaoHang-bio** [comment]
Hi,

The species I'm working with are from the same genus of Asteraceae plants, with an estimated divergence time of about 5 to 8 MYA. I don't believe they are very distantly related species.

Additionally, today I obtained some new TOGA results, all based on the same reference genome, with divergence times ranging between 2 to 10 MYA. Here's a summary of these results:

```
Sample	BUSCO (%)	Missing Genes
F7	64.0	6025
F3	81.0	5694
F4	26.4	15782
F6	48.7	15782
F8	38.3	5620
F1	51.3	13595
F5	41.6	6798
```
Given the divergence times and the BUSCO scores, along with the number of missing genes for each sample, I'm looking forward to your insights on these results.

Best regards,
Hang

**Comment by MichaelHiller** [comment]
OK, then divergence is not a problem. 
Maybe the chains are really the problem here. I can only guess. 
Can you visualize them in a genome browser (e.g. UCSC has the bigChain format; maybe IGV can also show chains)? This would allow you to see if missing genes that are actually present align in the chain and if introns/intergenic regions align.

Alternatively, maybe run the chains again with the parameters we set as default and then run TOGA again. If the results massively improve, then the parameters you picked where likely not appropriate.

**Comment by ZhaoHang-bio** [comment]
Hi,

I've observed a consistent issue where the alignment success rate for genes on chromosomes is significantly lower than that for genes on scaffolds in the TOGA output. This pattern correlates with the volume of data generated by lastz alignments, suggesting a potential limitation with lastz in handling longer sequences.

```
Number of mapping genes Prefix of geneID  ChrID
   5125 >FLBID01 Chr01
   3181 >FLBID02 Chr02
   3318 >FLBID03 Chr03
   2524 >FLBID04 Chr04
   3812 >FLBID05 Chr05
   2321 >FLBID06 Chr06
   2429 >FLBID07 Chr07
   2836 >FLBID08 Chr08
   3199 >FLBID09 Chr09
   2797 >FLBID10 Chr10
   2214 >FLBID11 Chr11
   3196 >FLBID12 Chr12
   2324 >FLBID13 Chr13
   2387 >FLBID14 Chr14
   1985 >FLBID15 Chr15
   2213 >FLBID16 Chr16
   2334 >FLBID17 Chr17
   2772 >FLBID18 Chr18
 455330 >FLBIDUn Scaffold
```
![image](https://github.com/hillerlab/make_lastz_chains/assets/58715839/d4799c15-86a9-45c8-9f70-23b6fc527162)


Is there a known limitation of lastz where it performs suboptimally with longer chromosomal sequences compared to shorter scaffold sequences? If so, what are the recommended workarounds or parameter adjustments to improve alignment success rates for longer sequences?

Additionally, I am attempting to generate alignment results using the make_lastz_chain tool, but I am encountering a significant number of failed subprocesses. I suspect this may be due to the program's runtime being forcibly set to 24 hours. How can this issue be addressed?
![image](https://github.com/hillerlab/make_lastz_chains/assets/58715839/b11c7771-7636-4e2b-aa83-de1351b110ab)
![image](https://github.com/hillerlab/make_lastz_chains/assets/58715839/b0cfd9cc-4cf5-48ae-b668-b764fa00d6be)

Do you think the characteristics of plant genomes, such as larger genome sizes and a higher proportion of repetitive sequences, could be contributing to this problem? Any advice on how to adjust the tool or process to accommodate these genomic features would be greatly appreciated.

I'm looking forward to your reply.

Best regards,
Hang

**Comment by MichaelHiller** [comment]
I believe longer chroms, simply have more genes (and then more are missing). 
lastz and chains has no problems with long chroms, as we split them up into smaller chunks for the lastz step anyway. 

But if many lastz jobs crash and did not finish, then this explains why chains are incomplete and why TOGA misses genes. 
Can you pls check the make_chains github if somebody already had this issue (if not, pls post. Bogdan can hopefully have a look)? 
And pls try to increase the runtime or use smaller chunk sizes to reduce runtimes. Your ref and query genomes was repeat-modeled and repeat-masked?

**Comment by ZhaoHang-bio** [comment]
Thank you for the response.

 I have posted a issue on the make_lastz_chains GitHub page seeking advice on how to extend the process time limits and recommendations for genome chunk sizes.

Additionally, the genomes files I used for alignment are all softmasked sequences by EDTA and RepeatMasker.

**Comment by ZhaoHang-bio** [comment]
I am still experiencing continuous timeouts with the `make_lastz_chains` program. I didn't find any option to extend the 24 hour time limit. Would switching from a soft-masked genome to a hard-masked genome potentially resolve these timeout issues? Additionally, I'm concerned about the impact this change might have on the TOGA results. Could you provide any insights into whether such a change in masking approach could affect the outcome?

**Comment by MichaelHiller** [comment]
Hardmasking is not ideal and would likely not reduce runtimes even more. 
If you can't extend the 24h time limit, pls try 2 things
* use windowMasker and add this to the RepeatMask (meaning more bases would now be lower case; WindowMasker better catches tandem and satellite repeats, e.g. in centromers)
* reduce the chunksize of the ref and query genome, each by half or more. This will give more jobs, but each is smaller and should run faster.

**Comment by ZhaoHang-bio** [comment]
Hi,

During my testing, I found that reducing the chunksize to "--seq1_chunk 5000000 --seq2_chunk 500000" prevents make_chains.py  from throwing any errors. However, this has severely impacted the performance, resulting in an extremely slow process. It has only completed about 0.6% in one days, which is an impractical execution speed for my needs.

Could you advise on any additional parameter adjustments that might accelerate the execution speed? Below is the command line I'm using:
`make_chains.py target query ${refgenome} ${quegenome} --pd chain_out -f --chaining_memory 100 --cluster_executor slurm --cluster_queue smp01,amd,low --keep_temp --seq1_chunk 5000000 --seq2_chunk 500000`

The output of make_chains.py is as follows:
`[a6/5e62c0] process > execute_jobs (3077) [ 0%] 3798 of 611694
executor > slurm (3898)
[a6/5e62c0] process > execute_jobs (3077) [ 0%] 3798 of 611694
executor > slurm (3899)
[91/c8f810] process > execute_jobs (3078) [ 0%] 3799 of 611694
executor > slurm (3900)
[ea/a0c4ee] process > execute_jobs (3079) [ 0%] 3800 of 611694`

I appreciate any help you can provide.

**Comment by MichaelHiller** [comment]
Slow speed is always an indication that repeatMasking did not work well. 
I was asking before, whether you generated a library of repeats individually for each genome with repeatModeler (we always do that) and then use this library with RepeatMasker to soft (not hard) mask?

Are your genomes on NCBI? If so, I can try to run it here.

**Comment by ZhaoHang-bio** [comment]
Thank you for your response. 

Yes, I have indeed generated a library of repeat sequences for each genome individually using the EDTA software, and subsequently performed soft-masking on each genome with RepeatMasker. 
My genomes are not sourced from NCBI.  I have already sent the test files for the two genomes to your email address for your evaluation. 

Looking forward to your feedback.

**Comment by MichaelHiller** [comment]
What I can say for sure is that -- as suspected -- your masking was insufficient. 

I ran your genomes and most alignment jobs did not finish for me either. 
So I used RepeatModeler (not EDTA) to create a lib and mask. 

For your query genome, I now mask 71.45% vs. 55.8% (EDTA).
For your reference genome, I now mask 78.93% vs. 61.1% (EDTA). 

Now almost all jobs run super fast. 
A few jobs are still running, likely because they contain tandem repeats which repeatMasker did not detect fully. 
I reduced chunk size to 20 Mb (ref) and 10 Mb (query), producing ~9000 jobs.

**Comment by MichaelHiller** [comment]
Here are the runtime stats. As I mentioned above, I split the genome into much smaller chunks as we normally do. 
Total time in finished jobs:       352329 sec   5872.1 min      97.9 h          4.1 days
Average job time:                39 sec      0.6 min       0.0 h          0.0 days
Longest finished job:          2428 sec     40.5 min       0.7 h          0.0 days      

The single job that took longer needed 36.5 h to finish. 

I'll send you the repeatMasker files, repeat lib and the chains.

Bottom line: It is recommended to use RepeatModeler and RepeatMasker. If that doesn't help, smaller chunk sizes and optionally adding windowmasking.
